package demoserver;

message EntitySearchRequest {
    required string dataset = 1;
    required string query = 2;
    optional int32 offset = 3;
    optional int32 count = 4;
}

message EntityDetailRequest {
    required string dataset = 1;
    repeated int32 id = 2;
}

message VDataDetail {
    required int32 id = 1;
    required string name = 2;
    repeated int32 communities = 3;
    required double score = 4;
}

message Edge {
    required int32 f = 1;
    required int32 l = 2;
}

message Outdata {
    required int32 id = 1;
    required double score = 2;
    required string name = 3;
    repeated int32 communities = 4;
    repeated VDataDetail neighbor = 5;
    repeated Edge edge = 6;
    required double pagerank = 7;
}

message AnalyzeRequest {
    required string path = 1;
    required string algorithm = 2;
}

message AnalyzeResponse {
    required string path = 1;
    repeated Outdata us = 2;
    required string algorithm = 3;
}

message Stat {
    required string type = 1;
    required int32 value = 2;
}

message RelatedEntity {
    required string type = 1;
    repeated int32 id = 2;
}

message DetailedEntity {
    required string title = 1;
    required int32 id = 2;
    optional string url = 3;
    optional string description = 4;
    repeated Stat stat = 5;
    optional string topics = 6;
    optional string imgurl = 7;
    optional int32 original_id = 8;
    repeated RelatedEntity related_entity = 9;
}

message EntitySearchResponse {
    required string query = 1;
    required int32 total_count = 2;
    repeated DetailedEntity entity = 3;
}

message Influence {
    required int32 id = 1;
    required int32 topic = 2;
    required double score = 3;
}

message InfluenceSearchResponse {
    required int32 entity_id = 1;
    repeated Influence influence = 2;
    repeated Influence influenced_by = 3;
}
